### PR TITLE
Thumbnail support for BC6 and BC7

### PR DESCRIPTION
**Pull request by thaeseler-amd**
## Description
Added thumbnail support for dds formats BC6 and BC7.

### CODE REVIEW

File: renderdoc/os/win32/win32_shellext.cpp [hunk 1]
```diff
@@ -449,6 +453,7 @@ struct RDCThumbnailProvider : public IThumbnailProvider, IInitializeWithStream
       decompressed.resize(thumbheight * thumbwidth * 4);
       unsigned char decompressedBlock[decompressedBlockMaxSize];
       unsigned char greenBlock[16];
+      unsigned short decompressedBC6[48];
```
Comment (baldurk) [inline renderdoc/os/win32/win32_shellext.cpp]: I'd prefer an explicitly sized format `uint16_t` here.

File: renderdoc/os/win32/win32_shellext.cpp [hunk 2]
```diff
@@ -487,9 +499,33 @@ struct RDCThumbnailProvider : public IThumbnailProvider, IInitializeWithStream
...
+                if(resourceType == ResourceFormatType::BC7 && isSRGB)
+                {
+                  for(int pixelInStride = 0; pixelInStride < 4; pixelInStride++)
+                  {
+                    int pixelIndex = pixelInStride * 4 + (i * 16);
+                    // get srgb value as float
+                    Vec4f srgb = Vec4f(decompPointer[pixelIndex] / 255.0f,
+                                       decompPointer[pixelIndex + 1] / 255.0f,
+                                       decompPointer[pixelIndex + 2] / 255.0f, 0.0f);
+                    Vec4f rgb = ConvertSRGBToLinear(srgb);
```
Comment (baldurk) [inline renderdoc/os/win32/win32_shellext.cpp]: Why are we doing an sRGB decode here? I assume that the output thumbnail is expected to be in sRGB space, and we're not decoding any of BC1-BC3 which are likely to be also stored in sRGB. Arguably we should check to see if the dds file is using the linear format and do a conversion _to_ sRGB, but as with the texture viewer I think it's more conventional to simply let explorer interpret the linear data as sRGB without conversion (as most viewers will do that).

File: renderdoc/os/win32/win32_shellext.cpp [hunk 3]
```diff
@@ -510,6 +546,30 @@ struct RDCThumbnailProvider : public IThumbnailProvider, IInitializeWithStream
...
+              case ResourceFormatType::BC6:
+                for(int pixelInStride = 0; pixelInStride < 4; pixelInStride++)
+                {
+                  // compute Pixel Index for the decompressedBC6 buffer. One block stide = 12
+                  // floats.
+                  int pixelIndex = pixelInStride * 3 + (i * 12);
+                  // decompressed BC6 block uses 16:16:16 color format. Convert to 8:8:8:8
+                  unsigned short r = decompressedBC6[pixelIndex];
+                  unsigned short g = decompressedBC6[pixelIndex + 1];
+                  unsigned short b = decompressedBC6[pixelIndex + 2];
```
Comment (baldurk) [inline renderdoc/os/win32/win32_shellext.cpp]: Same thing here please - prefer `uint16_t` sized format.

File: renderdoc/os/win32/win32_shellext.cpp [hunk 4]
```diff
@@ -510,6 +546,30 @@ struct RDCThumbnailProvider : public IThumbnailProvider, IInitializeWithStream
...
+                  // convert to half float
+                  float redF = ConvertFromHalf(r);
+                  float greenF = ConvertFromHalf(g);
+                  float blueF = ConvertFromHalf(b);
+                  // clamp to 0..1
+                  redF = RDCMAX(RDCMIN(1.0f, redF), 0.0f);
+                  greenF = RDCMAX(RDCMIN(1.0f, greenF), 0.0f);
+                  blueF = RDCMAX(RDCMIN(1.0f, blueF), 0.0f);
```
Comment (baldurk) [inline renderdoc/os/win32/win32_shellext.cpp]: There's `RDCCLAMP(x, 0.0f, 1.0f)` which is a little easier to read.

**Review by baldurk — CHANGES_REQUESTED** [review]

File: renderdoc/os/win32/win32_shellext.cpp (hunk 1 above)
Comment (thaeseler-amd) [reply renderdoc/os/win32/win32_shellext.cpp]: Changed to `uint16_t`.

File: renderdoc/os/win32/win32_shellext.cpp (hunk 3 above)
Comment (thaeseler-amd) [reply renderdoc/os/win32/win32_shellext.cpp]: Changed to `uint16_t`.

File: renderdoc/os/win32/win32_shellext.cpp (hunk 2 above)
Comment (thaeseler-amd) [reply renderdoc/os/win32/win32_shellext.cpp]: Removed the conversion from srgb to linear.

File: renderdoc/os/win32/win32_shellext.cpp (hunk 4 above)
Comment (thaeseler-amd) [reply renderdoc/os/win32/win32_shellext.cpp]: Thanks! Changed to `RDCCLAMP`.

**Review by baldurk — APPROVED** [review]

**Pull request merged**